### PR TITLE
WT-8170 Reduce complexity of timestamp usage assertion API code

### PR DESCRIPTION
**Pull request by keithbostic**
This is a first draft of the changes discussed in WT-8170 (and minor remaining comments from WT-8691), to consolidate and generally clean up the timestamp usage checking and the assert/verbose configurations. This merges all of the timestamp checking code into one function which is called at commit.

The `verbose=()` configuration is extended to work like `assert`, that is, `assert=(read_timestamps,write_timestamps=(on,off))`. This change requires a corresponding change in MDB Server, but what I'd prefer, assuming we go in this direction, is changing `write_timestamps` in both `verbose` and `assert` to be a boolean rather than having `on/off` choices.

The `assert=()` configuration is changed to always return an error (rather than assert in some cases). The `verbose=()` configuration is changed to always push out a message, but continue without error.

Because timestamp checking can fail, commit is now always 2-pass through the list of updates (previously, only fast-truncate processing required 2 passes), so there may be performance considerations in this change.

This change isn't done, but I want to make sure this change matches your vision of how assert/verbose configurations should work.

@bvpvamsikrishna, @kommiharibabu, I know this is of interest to you, @quchenhao, this may overlap with your work as well. Please take a look and let me know what you think. I don't need detailed review yet, but I need to know this is moving in the right direction. -- Thanks!

### CODE REVIEW

File: dist/api_data.py
Comment (kommiharibabu) [inline dist/api_data.py:71]: I like the change of bringing the read_timestamp also into the verbose. Just a thought is it good if we have a `read_timestamp_usage` similar to `write_timestamp_usage`?

File: src/include/txn_inline.h
Comment (kommiharibabu) [inline src/include/txn_inline.h:1282]: Later we are verifying the flags and raised the error message, can we remove the flag checks here?

File: src/txn/txn.c
Comment (kommiharibabu) [inline src/txn/txn.c:1489]: I think this check is already performed as part of `__wt_txn_set_timestamp`?

**Review by kommiharibabu — COMMENTED** [review]
The changes are good. IMO, they are going in the correct direction.

File: dist/api_data.py
Comment (keithbostic) [reply dist/api_data.py:71]: I thought about that too, but decided it didn't make a difference since `always` or `none` were the only configurations I could think of that would make sense for `read_timestamp_usage`, so it wouldn't make anything simpler to add that configuration string.

If you think there are more configurations that are interesting, I'm happy to talk through that change.

File: src/include/txn_inline.h
Comment (keithbostic) [reply src/include/txn_inline.h:1282]: Yes.

This check isn't correct anyway, and there's no real need for it. I did a similar (but hopefully correct), check in `__txn_timestamp_usage_check()`, because there's a log of additional work in that function we'd have to do even if we're not configured for assert or verbose, and that's not true here.

File: src/txn/txn.c
Comment (keithbostic) [reply src/txn/txn.c:1489]: I thought so too -- I was looking for that... :) 

I think the only place we check is `__wt_txn_set_durable_timestamp()`, which is the appropriate error message. So, shall I just remove this check entirely?

**Comment by keithbostic** [comment]
@kommiharibabu, @bvpvamsikrishna, I think WT-8710 (PR #7496) is ready for detailed review, please.

It passes MDB server tests as far as I can tell, with the exception that the `ordered` requirement for "*must use commit-TS once a commit-TS has been used*" doesn't pass. I believe that's fallout from making `ordered` behave the way we documented it behaves, that is, we documented `ordered` was a superset of `key_consistent`, but the code didn't implement that. Once we agree this PR is correct for WiredTiger, I'll talk it over with the MDB server group and figure out the right solution for that problem.

Here's the list of changes outside of the core rework of the TS-check functions:

* I extended the `verbose=` configuration to support `read_timestamp`, that is, `verbose=` and `assert=` are now the same.
*  The `verbose` option is no longer a verbose message configuration, even though it's called `verbose`, we output an error message when the rules are violated.
*  The `write_timestamp=` configurations are no longer `write_timestamp={on,off}`, `write_timestamp=` is now a boolean. This is incompatible with MDB server, but we're going to be incompatible with MDB server no matter what, and this is a cleaner fix. I'll figure out how to merge with MDB server once we approve this PR.
* The `WT_UPDATE.prev_durable_ts` field has been changed to be always available and checked if TS checks are configured. This increases the size of the `WT_UPDATE structure from 39B to 47B.
* We no longer have TS-check flags in the `WT_TXN` structure, we only use the flags in the `WT_DATA_HANDLE` structure.
* All the timestamp check messages include the string "*unexpected timestamp usage:*"
* We skip all TS-checks during recovery (we used to pick and choose the ones we skipped).
* I rewrote the TS-check tests (test_assert01, 02, 03, 04 and 05), as test_timestamp26 -- I should rewrite test_assert06 too, but I haven't yet.

The core-TS check function rework is all in `txn.c`, and consisted of merging `__txn_commit_timestamps_usage_check()` and `__txn_commit_timestamps_assert_standalone()` into a single function, `__txn_timestamp_usage_check()`.

I then replaced the previous calls in `__txn_resolve_prepared_op()` and `__wt_txn_commit()` with calls to the new function. Additionally, `__wt_txn_commit()` now does two passes through the modifications array (because it needs to be able to rollback the transaction after `__txn_timestamp_usage_check()` fails). This work is where I have the most concerns, specifically:

1.  Did I incorrectly merge the two functions such that we're not doing the right set of checks any more?
2. Is it  correct for `__txn_resolve_prepared_op()` to call `__txn_commit_timestamps_usage_check()`?
3. Is the re-ordering of `__wt_txn_commit()` correct?

-- Thanks!

**Review by kommiharibabu — APPROVED** [review]
LGTM

**Comment by keithbostic** [comment]
@dgottlieb, hopefully this is the branch you want. :)

As we discussed `verbose=` is completely ignored and undocumented. If `assert=` is configured for read or write timestamps, error checking happens and error messages are logged. If `assert=` is configured in a DIAGNOSTIC build, we drop core.

I'm going to hold off on general review until you're happy with it, please let me know what I can do to help!

**Comment by keithbostic** [comment]
@sueloverso, commit 71ae183 has the changes we discussed: (1) remove `WT_SESSION_NO_LOGGING` in favor of checking `WT_CONN_RECOVERING`, and (2) review the "are we writing log records" checks in `__wt_txn_log_op()` and `__wt_txn_ts_log()`.

Would you please review the change?

It took me awhile to figure out why, but the checks in those functions were actually correct: we don't log timestamp operations in anything other then debug mode, and since we may or may not have a BTREE handle at that time, the correct check is if we are in DEBUG mode. (I think there are bugs lurking in there, because I saw a core dump when I accidentally called the timestamp-log function without a valid BTREE handle, but I lost that code change and it didn't immediately reproduce. But, yeah, I think there are assumptions in there about when DEBUG mode is configured.)

All that said, I still reworked and re-ordered the "are we writing log records" tests because it's a function call to figure out we're not going to write a log record, and for MDB server, writing a log record on update is an exception, not the rule.

File: dist/api_data.py
Comment (sueloverso) [inline dist/api_data.py:77]: Since 'verbose' is no longer supported you should get rid of it in this description.

File: dist/api_data.py
Comment (sueloverso) [inline dist/api_data.py:85]: Related to 'verbose' it seems the phrase 'log message' in this description is also meant for the 'verbose' setting.

File: src/include/txn.h
Comment (sueloverso) [inline src/include/txn.h:328]: Trying to understand what changed here? I see you removed the `WT_TXN_TS_WRITE*` lines. But why did removing those modify all the earlier flags (which have the same values as before)? That is unexpected.

File: test/suite/test_timestamp26.py
Comment (sueloverso) [inline test/suite/test_timestamp26.py:331]: What is the point of leaving this in here? Is it supposed to be made real before merge and was forgotten?

File: test/suite/test_timestamp26.py
Comment (sueloverso) [inline test/suite/test_timestamp26.py:37]: Initially I was fairly alarmed at the number of tests removed. It seems this like it (mostly? completely?) encompasses the content of those. I did not compare detailed content however.

File: test/suite/test_assert06.py
Comment (sueloverso) [inline test/suite/test_assert06.py:36]: It seems strange to leave this one test in the `test_assert06.py` when all the rest were deleted and moved to the other test. There is a `test_assert07.py` that seems mis-categorized. Consider renaming these two.

**Review by sueloverso — APPROVED** [review]
LGTM.

File: dist/api_data.py
Comment (keithbostic) [reply dist/api_data.py]: Fixed in c83401348.

File: dist/api_data.py
Comment (keithbostic) [reply dist/api_data.py]: `assert=` configurations, outside of `HAVE_DIAGNOSTIC` builds output error messages, so I think this is OK.

File: test/suite/test_timestamp26.py
Comment (keithbostic) [reply test/suite/test_timestamp26.py]: Nice catch! I missed one of the tests when I was updating them. Fixed in 6487e0adb.

File: test/suite/test_timestamp26.py
Comment (keithbostic) [reply test/suite/test_timestamp26.py:37]: They were largely rewritten, comparing in detail would be almost impossible.

File: test/suite/test_assert06.py
Comment (keithbostic) [reply test/suite/test_assert06.py:36]: Honestly, I ran out of steam after rewriting all of the others. It was pretty painful. :(

File: src/include/txn.h
Comment (sueloverso) [reply src/include/txn.h:328]: Right! Thanks. I wasn't seeing the one-less-zero. I thought I was missing some kind of whitespace change.

**Review by agorrod — COMMENTED** [review]
Thanks @keithbostic this change generally seems good to me.

Could the tests be updated to look for messages in assert mode instead of being removed?

**Comment by keithbostic** [comment]
> Could the tests be updated to look for messages in assert mode instead of being removed? FWIW: I suspect it's because debug builds will crash, but wanted to ask the question..

Yes we can. That's what I've done, by limiting those tests to builds without HAVE_DIAGNOSTIC. So, we will run the usual tests, just on fewer builds/platforms.

**Comment by keithbostic** [comment]
@sueloverso, @dgottlieb, @kommiharibabu, @agorrod: I think that WT-8170 (#7496) is ready to go, and I'll merge it once SERVER-63694 is in.

There have been a bunch of fundamental changes after the PR was first created, based on MDB Server needs, so I wanted to give everybody one last chance to object before it goes in.

A couple of items:

* The truncation logging code (https://github.com/wiredtiger/wiredtiger/blob/wt-8170-timestamp-simplify/src/btree/bt_cursor.c#L1818-L1869), has changed. We used to only check if database logging was enabled before calling the logging functions for truncate, which I’m pretty sure was wrong — we shouldn’t do that work if the table isn’t logged -- but it's a real change.
* I'm releasing the transaction snapshot a lot earlier in `__wt_txn_commit()`. I couldn't think of any reason not to do that, and it should release resources earlier vs. later.
* There are always 2 passes through the list of updates in `__wt_txn_commit()`. We could get it down to one if the transaction isn't a prepare, at the cost of some complexity. I'm assuming there generally aren't that many updates to review, and that it won't matter.

Anyway, if you're happy, click "yes". :)

**Review by sueloverso — APPROVED** [review]
LGTM. I have a couple very minor questions.

**Comment by dgottlieb** [comment]
Dumb question re:
```
I'm releasing the transaction snapshot a lot earlier in __wt_txn_commit(). I couldn't think of any reason not to do that, and it should release resources earlier vs. later.
```
When do we release the snapshot in relation to walking a txn's linked list of WT_UPDATE* entries? How does WT know those pointer dereferences are valid?

Just curious here, not suggesting the change breaks anything I'm asking about. The statement just made me realize that there's more nuance to update structure lifetimes than my existing more coarse-grained model from the application's perspective.

File: src/include/txn_inline.h
Comment (keithbostic) [reply src/include/txn_inline.h:1299]: @dgottlieb has run full patch builds, I believe we're OK with this change.

File: src/txn/txn.c
Comment (keithbostic) [reply src/txn/txn.c]: Fixed in 2f6b9ffea.

**Comment by keithbostic** [comment]
> When do we release the snapshot in relation to walking a txn's linked list of WT_UPDATE* entries? How does WT know those pointer dereferences are valid?

In the current model, the update structure must stay in memory until the transaction completes, but that's separate from the visibility rules requiring other updates to stay in memory until the transaction completes. In other words, releasing the snapshot informs the system we won't be reading data the engine might be keeping in memory, just in case we do more reads.

File: src/include/txn_inline.h
Comment (dgottlieb) [reply src/include/txn_inline.h:1299]: Thanks for flagging this @sueloverso. I think this one is fine. Specifically related to read timestamp assertions -- we don't use those. More generally, it is a little hard to know which WT strings MDB testing depends on. I believe perf testing at some point in the past got the idea to check for WT warnings and interpreting their existence as a test failure. So I appreciate calling this one out. Even if I don't know how to exhaustively check who might be depending on it.

File: src/txn/txn.c
Comment (dgottlieb) [inline src/txn/txn.c:869]: I struggle to read basic predicates, but I believe this means in-memory always checks timestamp usage, regardless of whether the table is logged?

Based on our recent findings with in-memory I presume you'll want to remove the in memory check here? We're just digging a hole where one day (maybe today?) inmemory will start failing because it enforces more constraints than durable WT.

File: src/txn/txn.c
Comment (dgottlieb) [inline src/txn/txn.c:1504]: Ah my question!

So it seems that releasing the snapshot can cause WT_UPDATE pointers to go invalid? This comment refers to cursors, so I can see how this keeps cursors valid. But I wouldn't expect all of the updates in the `mods` list to have a corresponding cursor.

That said, the old code was releasing the snapshot before walking the `mod` list. Definitely curious to learn more about this.

**Review by dgottlieb — APPROVED** [review]
I did hand check this produces core dumps as requested in diagnostics mode.

**Review by dgottlieb — COMMENTED** [review]
Oops, forgot I had an open question re: the inmemory flag for timestamp usage checking.

File: src/txn/txn.c
Comment (keithbostic) [reply src/txn/txn.c:869]: Yes, exactly; this is one of the places that will change based on WT-8449. Hopefully this one goes in first, but regardless, the merge won't be too bad.

File: src/txn/txn.c
Comment (keithbostic) [reply src/txn/txn.c:1504]: Yes; a page with unresolved updates cannot be evicted, that's a separate thing. This isn't exactly my area of expertise, but in general, the snapshot is about making sure future reads the transaction might make are guaranteed to not be overtaken by eviction (or the history store), not preventing existing writes from being evicted. And, the sooner we can release the snapshot, the sooner other threads can quit worrying about future reads from this transaction.

**Comment by keithbostic** [comment]
> Oops, forgot I had an open question re: the inmemory flag for timestamp usage checking.

I'm expecting to resolve that in WT-8449. As soon as SERVER-63694 merges, I'll merge this, and then immediately fix the WT-8601 problem in WT-8449.

File: src/txn/txn.c
Comment (dgottlieb) [reply src/txn/txn.c:1504]: Yeah releasing the snapshot to free resources of other documents makes sense. But I think you answered my question by stating how a page keeps track if there are unresolved commits on it.

File: test/suite/test_assert01.py
Comment (agorrod) [inline test/suite/test_assert01.py:1]: @keithbostic - it looks from the view here as though test_assert01 through 06 are being deleted, not updated to only operate when not in diagnostic mode.

Comment (keithbostic) [reply test/suite/test_assert01.py:1]: @agorrod
> it looks from the view here as though test_assert01 through 06 are being deleted, not updated to only operate when not in diagnostic mode.

I rewrote those tests pretty much from scratch, so I moved them into a new file, `test_timestamp26.py`.

Comment (agorrod) [reply test/suite/test_assert01.py:1]: Got it - thanks.

**Review by agorrod — APPROVED** [review]

**Pull request merged**
